File: aeron-archive/src/main/java/io/aeron/archive/client/ControlResponsePoller.java
/*
 * Copyright 2014-2025 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.aeron.archive.client;

import io.aeron.Aeron;
import io.aeron.ControlledFragmentAssembler;
import io.aeron.Subscription;
import io.aeron.archive.codecs.*;
import io.aeron.logbuffer.ControlledFragmentHandler;
import io.aeron.logbuffer.Header;
import org.agrona.DirectBuffer;
import org.agrona.SemanticVersion;

import java.util.Objects;

/**
 * Encapsulate the polling and decoding of archive control protocol response messages.
 */
public final class ControlResponsePoller
{
    /**
     * Limit to apply when polling response messages.
     */
    public static final int FRAGMENT_LIMIT = 10;

    private final MessageHeaderDecoder messageHeaderDecoder = new MessageHeaderDecoder();
    private final ControlResponseDecoder controlResponseDecoder = new ControlResponseDecoder();
    private final ChallengeDecoder challengeDecoder = new ChallengeDecoder();
    private final RecordingSignalEventDecoder recordingSignalEventDecoder = new RecordingSignalEventDecoder();

    private final Subscription subscription;
    private final ControlledFragmentAssembler fragmentAssembler = new ControlledFragmentAssembler(this::onFragment);
    private final int fragmentLimit;

    private long controlSessionId = Aeron.NULL_VALUE;
    private long correlationId = Aeron.NULL_VALUE;
    private long relevantId = Aeron.NULL_VALUE;
    private int templateId = Aeron.NULL_VALUE;
    private int version = 0;
    private ControlResponseCode code = null;
    private String errorMessage = null;
    private long recordingId = Aeron.NULL_VALUE;
    private long subscriptionId = Aeron.NULL_VALUE;
    private long position = Aeron.NULL_VALUE;
    private RecordingSignal recordingSignal = null;
    private byte[] encodedChallenge = null;
    private boolean isPollComplete = false;

    /**
     * Create a poller for a given subscription to an archive for control response messages with a default
     * fragment limit for polling as {@link #FRAGMENT_LIMIT}.
     *
     * @param subscription to poll for new events.
     */
    public ControlResponsePoller(final Subscription subscription)
    {
        this(subscription, FRAGMENT_LIMIT);
    }

    /**
     * Create a poller for a given subscription to an archive for control response messages.
     *
     * @param subscription  to poll for new events.
     * @param fragmentLimit to apply when polling.
     */
    public ControlResponsePoller(final Subscription subscription, final int fragmentLimit)
    {
        this.subscription = Objects.requireNonNull(subscription);
        this.fragmentLimit = fragmentLimit;
    }

    /**
     * Get the {@link Subscription} used for polling responses.
     *
     * @return the {@link Subscription} used for polling responses.
     */
    public Subscription subscription()
    {
        return subscription;
    }

    /**
     * Poll for control response events.
     *
     * @return the number of fragments read during the operation. Zero if no events are available.
     */
    public int poll()
    {
        if (isPollComplete)
        {
            controlSessionId = Aeron.NULL_VALUE;
            correlationId = Aeron.NULL_VALUE;
            relevantId = Aeron.NULL_VALUE;
            templateId = Aeron.NULL_VALUE;
            version = 0;
            code = null;
            errorMessage = null;
            recordingId = Aeron.NULL_VALUE;
            subscriptionId = Aeron.NULL_VALUE;
            position = Aeron.NULL_VALUE;
            encodedChallenge = null;
            recordingSignal = null;
            isPollComplete = false;
        }

        return subscription.controlledPoll(fragmentAssembler, fragmentLimit);
    }

    /**
     * SBE template id of polled message or {@link Aeron#NULL_VALUE} if poll returned nothing.
     *
     * @return SBE template id of polled message or {@link Aeron#NULL_VALUE} if poll returned nothing.
     */
    public int templateId()
    {
        return templateId;
    }

    /**
     * Control session id of polled message or {@link Aeron#NULL_VALUE} if poll returned nothing.
     *
     * @return control session id of polled message or {@link Aeron#NULL_VALUE} if poll returned nothing.
     */
    public long controlSessionId()
    {
        return controlSessionId;
    }

    /**
     * Correlation id of the message or {@link Aeron#NULL_VALUE} if poll returned nothing.
     *
     * @return correlation id of polled message or {@link Aeron#NULL_VALUE} if poll returned nothing.
     */
    public long correlationId()
    {
        return correlationId;
    }

    /**
     * Get the relevant id returned with the response, e.g. replay session id.
     *
     * @return the relevant id returned with the response.
     */
    public long relevantId()
    {
        return relevantId;
    }

    /**
     * Recording id of polled {@link RecordingSignal} or {@link Aeron#NULL_VALUE} if poll returned nothing.
     *
     * @return recording id of polled {@link RecordingSignal} or {@link Aeron#NULL_VALUE} if poll returned nothing.
     */
    public long recordingId()
    {
        return recordingId;
    }

    /**
     * Subscription id of polled {@link RecordingSignal} or {@link Aeron#NULL_VALUE} if poll returned nothing.
     *
     * @return subscription id of polled {@link RecordingSignal} or {@link Aeron#NULL_VALUE} if poll returned nothing.
     */
    public long subscriptionId()
    {
        return subscriptionId;
    }

    /**
     * Position of polled {@link RecordingSignal} or {@link Aeron#NULL_VALUE} if poll returned nothing.
     *
     * @return position of polled {@link RecordingSignal} or {@link Aeron#NULL_VALUE} if poll returned nothing.
     */
    public long position()
    {
        return position;
    }

    /**
     * Enum of polled {@link RecordingSignal} or null if poll returned nothing.
     *
     * @return enum of polled {@link RecordingSignal} or null if poll returned nothing.
     */
    public RecordingSignal recordingSignal()
    {
        return recordingSignal;
    }

    /**
     * Version response from the server in semantic version form.
     *
     * @return response from the server in semantic version form.
     */
    public int version()
    {
        return version;
    }

    /**
     * Has the last polling action received a complete message?
     *
     * @return true if the last polling action received a complete message?
     */
    public boolean isPollComplete()
    {
        return isPollComplete;
    }

    /**
     * Get the response code of the last response.
     *
     * @return the response code of the last response.
     */
    public ControlResponseCode code()
    {
        return code;
    }

    /**
     * Get the error message of the response.
     *
     * @return the error message of the response.
     */
    public String errorMessage()
    {
        return errorMessage;
    }

    /**
     * Was the last polling action received a challenge message?
     *
     * @return true if the last polling action received was a challenge message, false if not.
     */
    public boolean wasChallenged()
    {
        return null != encodedChallenge;
    }

    /**
     * Get the encoded challenge of the last challenge.
     *
     * @return the encoded challenge of the last challenge.
     */
    public byte[] encodedChallenge()
    {
        return encodedChallenge;
    }

    ControlledFragmentAssembler.Action onFragment(
        final DirectBuffer buffer, final int offset, final int length, final Header header)
    {
        if (isPollComplete)
        {
            return ControlledFragmentHandler.Action.ABORT;
        }

        messageHeaderDecoder.wrap(buffer, offset);

        final int schemaId = messageHeaderDecoder.schemaId();
        if (schemaId != MessageHeaderDecoder.SCHEMA_ID)
        {
            throw new ArchiveException("expected schemaId=" + MessageHeaderDecoder.SCHEMA_ID + ", actual=" + schemaId);
        }

        templateId = messageHeaderDecoder.templateId();
        switch (templateId)
        {
            case ControlResponseDecoder.TEMPLATE_ID:
            {
                controlResponseDecoder.wrap(
                    buffer,
                    offset + MessageHeaderEncoder.ENCODED_LENGTH,
                    messageHeaderDecoder.blockLength(),
                    messageHeaderDecoder.version());

                controlSessionId = controlResponseDecoder.controlSessionId();
                correlationId = controlResponseDecoder.correlationId();
                relevantId = controlResponseDecoder.relevantId();
                code = controlResponseDecoder.code();
                version = controlResponseDecoder.version();
                errorMessage = controlResponseDecoder.errorMessage();
                isPollComplete = true;

                return ControlledFragmentHandler.Action.BREAK;
            }

            case ChallengeDecoder.TEMPLATE_ID:
            {
                challengeDecoder.wrap(
                    buffer,
                    offset + MessageHeaderEncoder.ENCODED_LENGTH,
                    messageHeaderDecoder.blockLength(),
                    messageHeaderDecoder.version());

                controlSessionId = challengeDecoder.controlSessionId();
                correlationId = challengeDecoder.correlationId();
                relevantId = Aeron.NULL_VALUE;
                code = ControlResponseCode.NULL_VAL;
                version = challengeDecoder.version();
                errorMessage = "";

                final int encodedChallengeLength = challengeDecoder.encodedChallengeLength();
                encodedChallenge = new byte[encodedChallengeLength];
                challengeDecoder.getEncodedChallenge(encodedChallenge, 0, encodedChallengeLength);

                isPollComplete = true;

                return ControlledFragmentHandler.Action.BREAK;
            }

            case RecordingSignalEventDecoder.TEMPLATE_ID:
            {
                recordingSignalEventDecoder.wrap(
                    buffer,
                    offset + MessageHeaderDecoder.ENCODED_LENGTH,
                    messageHeaderDecoder.blockLength(),
                    messageHeaderDecoder.version());

                controlSessionId = recordingSignalEventDecoder.controlSessionId();
                correlationId = recordingSignalEventDecoder.correlationId();
                recordingId = recordingSignalEventDecoder.recordingId();
                subscriptionId = recordingSignalEventDecoder.subscriptionId();
                position = recordingSignalEventDecoder.position();
                recordingSignal = recordingSignalEventDecoder.signal();

                isPollComplete = true;

                return ControlledFragmentHandler.Action.BREAK;
            }
        }

        return ControlledFragmentHandler.Action.CONTINUE;
    }

    /**
     * {@inheritDoc}
     */
    public String toString()
    {
        return "ControlResponsePoller{" +
            "templateId=" + templateId +
            ", controlSessionId=" + controlSessionId +
            ", correlationId=" + correlationId +
            ", relevantId=" + relevantId +
            ", recordingId=" + recordingId +
            ", subscriptionId=" + subscriptionId +
            ", position=" + position +
            ", recordingSignal=" + recordingSignal +
            ", code=" + code +
            ", version=" + SemanticVersion.toString(version) +
            ", errorMessage='" + errorMessage + '\'' +
            ", isPollComplete=" + isPollComplete +
            '}';
    }
}

File: aeron-archive/src/main/cpp/client/ControlResponsePoller.cpp
/*
 * Copyright 2014-2025 Real Logic Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ControlResponsePoller.h"
#include "ArchiveException.h"
#include "aeron_archive_client/MessageHeader.h"
#include "aeron_archive_client/ControlResponse.h"
#include "aeron_archive_client/Challenge.h"
#include "aeron_archive_client/RecordingSignalEvent.h"

using namespace aeron;
using namespace aeron::archive::client;

static aeron::controlled_poll_fragment_handler_t controlHandler(ControlResponsePoller &poller)
{
    return
        [&](AtomicBuffer &buffer, util::index_t offset, util::index_t length, Header &header)
        {
            return poller.onFragment(buffer, offset, length, header);
        };
}

ControlResponsePoller::ControlResponsePoller(std::shared_ptr<Subscription> subscription, int fragmentLimit) :
    m_fragmentAssembler(controlHandler(*this)),
    m_fragmentHandler(m_fragmentAssembler.handler()),
    m_subscription(std::move(subscription)),
    m_fragmentLimit(fragmentLimit)
{
}

ControlledPollAction ControlResponsePoller::onFragment(
    AtomicBuffer &buffer, util::index_t offset, util::index_t length, Header &)
{
    if (m_isPollComplete)
    {
        return ControlledPollAction::ABORT;
    }

    MessageHeader msgHeader(
        buffer.sbeData() + offset,
        static_cast<std::uint64_t>(length),
        MessageHeader::sbeSchemaVersion());

    const std::uint16_t schemaId = msgHeader.schemaId();
    if (schemaId != MessageHeader::sbeSchemaId())
    {
        throw ArchiveException(
            "expected schemaId=" + std::to_string(MessageHeader::sbeSchemaId()) +
            ", actual=" + std::to_string(schemaId),
            SOURCEINFO);
    }

    switch (msgHeader.templateId())
    {
        case ControlResponse::SBE_TEMPLATE_ID:
        {
            ControlResponse response(
                buffer.sbeData() + offset + MessageHeader::encodedLength(),
                static_cast<std::uint64_t>(length) - MessageHeader::encodedLength(),
                msgHeader.blockLength(),
                msgHeader.version());

            m_controlSessionId = response.controlSessionId();
            m_correlationId = response.correlationId();
            m_relevantId = response.relevantId();
            m_version = response.version();

            ControlResponseCode::Value code = response.code();
            m_codeValue = code;
            m_isCodeError = ControlResponseCode::Value::ERROR == code;
            m_isCodeOk = ControlResponseCode::Value::OK == code;

            m_errorMessage = response.getErrorMessageAsString();

            m_isControlResponse = true;
            m_isPollComplete = true;

            return ControlledPollAction::BREAK;
        }

        case Challenge::SBE_TEMPLATE_ID:
        {
            Challenge challenge(
                buffer.sbeData() + offset + MessageHeader::encodedLength(),
                static_cast<std::uint64_t>(length) - MessageHeader::encodedLength(),
                msgHeader.blockLength(),
                msgHeader.version());

            m_controlSessionId = challenge.controlSessionId();
            m_correlationId = challenge.correlationId();
            m_relevantId = aeron::NULL_VALUE;
            m_version = challenge.version();

            m_codeValue = ControlResponseCode::NULL_VALUE;
            m_isCodeError = false;
            m_isCodeOk = false;
            m_errorMessage = "";

            const std::uint32_t encodedChallengeLength = challenge.encodedChallengeLength();
            char *encodedBuffer = new char[encodedChallengeLength];
            challenge.getEncodedChallenge(encodedBuffer, encodedChallengeLength);

            m_encodedChallenge.first = encodedBuffer;
            m_encodedChallenge.second = encodedChallengeLength;

            m_isControlResponse = false;
            m_wasChallenged = true;
            m_isPollComplete = true;

            return ControlledPollAction::BREAK;
        }

        case RecordingSignalEvent::SBE_TEMPLATE_ID:
        {
            RecordingSignalEvent recordingSignalEvent(
                buffer.sbeData() + offset + MessageHeader::encodedLength(),
                static_cast<std::uint64_t>(length) - MessageHeader::encodedLength(),
                msgHeader.blockLength(),
                msgHeader.version());

            m_controlSessionId = recordingSignalEvent.controlSessionId();
            m_correlationId = recordingSignalEvent.correlationId();
            m_recordingId = recordingSignalEvent.recordingId();
            m_subscriptionId = recordingSignalEvent.subscriptionId();
            m_position = recordingSignalEvent.position();
            m_recordingSignalCode = recordingSignalEvent.signalRaw();

            m_isRecordingSignal = true;
            m_isPollComplete = true;

            return ControlledPollAction::BREAK;
        }
    }


    return ControlledPollAction::CONTINUE;
}
